### DISCUSSION TITLE
No rates present in response headers

### DISCUSSION THREAD

**Discussion by g-normand** [hozn/stravalib#593] [Q&A]
Hello,
I'm using Stravalib 1.6 with python 3.9.

My code is something like:
```
        self.client = Client(rate_limiter=DefaultRateLimiter(priority='medium'))
        self.client.access_token = access_token
        ...
        self.check_token()
        return self.client.upload_activity(*args, **kwargs)
```
But I get "No rates present in response headers" as a warning.
It looks like it changed on the 1.6 version (https://github.com/stravalib/stravalib/pull/446/ ), when I was using the 1.3.3 version I didn't had that kind of warnings.

What is the problem there?
Thanks,
Guillaume

**Comment by jsamoocha** [comment]
Hi!

I tried something similar with 1.6 and py 3.9, and DEBUG logging level. I see no warnings and this is the complete (anonymized) logging output:

```
INFO:stravalib.protocol.ApiV3:GET 'https://www.strava.com/api/v3/athlete' with params {}
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): www.strava.com:443
DEBUG:urllib3.connectionpool:https://www.strava.com:443 "GET /api/v3/athlete?access_token=XXXXXX HTTP/11" 200 None
DEBUG:stravalib.util.limiter.SleepingRateLimitRule:Throttling based on rates: RequestRate(short_usage=1, long_usage=47, short_limit=300, long_limit=3000)
```

I cannot easily test it for the `upload_activity()` endpoint as this token was linked to a real Strava account.

Are you familiar with the [Strava Developer Program](https://communityhub.strava.com/t5/developer-knowledge-base/our-developer-program/ta-p/8849), and have you applied for your app?

**Reply by lwasser** [reply]
yes @g-normand I second what @jsamoocha mentioned above. Can you also please try to troubleshoot that your token is indeed valid outside of whatever workshop you are using? For example, does client.get_athlete() process correctly? This will allow you to rule out whether it is indeed an authentication error which I think we both suspect is the case. I was able to recreate the above error with a token that wasn't refreshed. It also makes sense that having an app that isn't setup quite right causing a similar issue.

**Reply by g-normand** [reply]
Ok, I found the problem : 

My code is : 
```
    def check_token(self):
        """
        It's up to us to check if the token is still valid.
        """
        if now() > self.expires_at:
            self.client = Client(rate_limiter=DefaultRateLimiter(priority='medium'))
            refresh_response = self.client.refresh_access_token(
                client_id=settings.STRAVA_CLIENT_ID,
                client_secret=settings.STRAVA_CLIENT_SECRET,
                refresh_token=self.refresh_token
            )
            self.access_token = self.client.access_token = refresh_response['access_token']
            self.refresh_token = refresh_response['refresh_token']
            self.expires_at = make_aware(
                datetime.datetime.utcfromtimestamp(refresh_response['expires_at'])
            )
            self.save()
```

if datetime.now > expires_at, it will ask for a refresh_access_token, but I will get the "No rates present in response headers" warning. Is there any other way to process? Thanks.

**Reply by lwasser** [reply]
hey @g-normand thanks for digging into this further. So I just ran through this myself. I do see that message! BUT it also does refresh the token properly. I'm following what [we wrote in our tutorial here](https://stravalib.readthedocs.io/en/latest/get-started/how-to-get-strava-data-python.html#step-4-use-and-refresh-your-strava-api-token) . So my question for you is, when you see that message are you still able to authenticate because I am. 

I see that message, but I am still able to access Strava data via the API after the message occurs. So as you mention above it's a warning but not one that actually impacts my local workflow. 

If that is the case, then I think we just need to make some adjustments in how stravalib provides feedback in that token exchange process. Also please note that we have [this issue open](https://github.com/stravalib/stravalib/issues/585) where we do plan to add a decorator that will handle the refresh for you. 

but first, please confirm that the message is just a message and not actually blocking your workflow. If you can confirm that I suggest we add a comment to our existing issue to handle the messaging that is provided as an output when we create the decorator. 

BTW I'm testing on stravalib 2.1 the latest version.

**Reply by g-normand** [reply]
Thanks @lwasser for the answer. It's just a warning, as you said, so it doesn't impact the workflow

**Reply by woodmicha** [reply]
Is there a way to silence the warnings?
